internal/cli/auth/login.go

```diff
@@ -291,43 +241,35 @@ func (opts *LoginOpts) oauthFlow(ctx context.Context) error {
 	}
 }
 
-func (opts *LoginOpts) shouldRetryAuthenticate(err error) (retry bool, errSurvey error) {
+func shouldRetryAuthenticate(err error, p survey.Prompt) (retry bool, errSurvey error) {
 	if err == nil || !auth.IsTimeoutErr(err) {
 		return false, nil
 	}
-
-	return opts.regenerateCodePrompt.confirm()
-}
-
-func hasUserProgrammaticKeys() bool {
-	return config.PublicAPIKey() != "" && config.PrivateAPIKey() != ""
+	err = telemetry.TrackAskOne(p, &retry)
+	return retry, err
 }
 
-func loginPreRun(ctx context.Context) error {
-	if hasUserProgrammaticKeys() {
-		msg := fmt.Sprintf(alreadyAuthenticatedError, config.PublicAPIKey())
-		return fmt.Errorf(`%s
-
-%s`, msg, loginWithProfileMsg)
+func newRegenerationPrompt() survey.Prompt {
+	return &survey.Confirm{
+		Message: "Your one-time verification code is expired. Would you like to generate a new one?",
+		Default: true,
 	}
-
-	if account, err := AccountWithAccessToken(); err == nil {
-		if err := cli.RefreshToken(ctx); err == nil && validate.Token() == nil {
-			msg := fmt.Sprintf(alreadyAuthenticatedEmailError, account)
-			return fmt.Errorf(`%s
-
-%s`, msg, logoutToLoginAccountMsg)
-		}
-	}
-	return nil
 }
 
-func (opts *LoginOpts) PreRun() error {
-	opts.config = config.Default()
-	if config.OpsManagerURL() != "" {
-		opts.OpsManagerURL = config.OpsManagerURL()
+func (opts *LoginOpts) LoginPreRun(ctx context.Context, c LoginConfig) func() error {
+	return func() error {
+		opts.config = c
+		if config.OpsManagerURL() != "" {
+			opts.OpsManagerURL = config.OpsManagerURL()
+		}
+		// ignore expired tokens since logging in
+		if err := opts.RefreshAccessToken(ctx); err != nil && !errors.Is(err, cli.ErrInvalidRefreshToken) {
+			return err
+		}
+		// clean up any expired tokens
+		opts.config.Set(config.AccessTokenField, "")
```

Comment:
this happens before we run validate.NoAccessToken validation (the new line 297) so the user will be allowed to re-login without doing first an `auth logout`. Is this intentional?

```
opts.InitFlow(config.Default()),
				opts.LoginPreRun(cmd.Context(), config.Default()),
				validate.NoAPIKeys,
				validate.NoAccessToken,
```